### ISSUE TITLE
Add option to choose the source/target nodes based on given probabilities (Approximate current flow betweenness centrality)

### ISSUE THREAD

**Issue by cam4ani** [networkx/networkx#6342]
### Discussed in https://github.com/networkx/networkx/discussions/6314

<div type='discussions-op-text'>

<sup>Originally posted by **cam4ani** January  3, 2023</sup>
Hello,

I am trying to use the approximate_current_flow_betweenness_centrality() function on a very large graph (millions of nodes and edges), but would like to choose the source and target nodes (i.e. the sample node pairs to use for approximation). For this, I replaced from the networkX code (https://networkx.org/documentation/stable/_modules/networkx/algorithms/centrality/current_flow_betweenness.html#approximate_current_flow_betweenness_centrality):
```
for _ in range(k):
     s, t = pair = seed.sample(range(n), 2)
```

with: 
```
#based on the seed given by the user, define a list of seeds used to select the target/source nodes
random.seed(seed)
li_randomseeds = random.sample(range(k*2), k=k)
for _ in range(k):
     dico_init_new = dict(zip(ordering, range(n)))
     random.seed(li_randomseeds[_])
     t,s = random.sample(df['ID_graph'].tolist(), counts=df['weight'].tolist(), k=2)
     #convert from the "G" labels to the new labels (from graph "H" defined in this function, on a previous line)
     t = dico_init_new[t]
     s = dico_init_new[s]
     pair = t,s
```
where for now "df" is a new argument to the function, which is simply a data frame of two columns where: "weight" represent how likely a node is to be chosen as a source/target, and  "ID_graph" is the initial ID of the graph "G" (the dataframe could be changed to another more suited data structure).  I tried it on a small graph and it seems to give the expected results, but I may be missing an important point as I do not understand the code fully. So if someone could tell me if this is the correct way to proceed or not it would be greatly appreciated.

Many thanks for your time, and also for this incredible package!</div>

**Comment by dschult** [comment]
I don't think this should become part of the current function within NetworkX because selecting the nodes randomly is an integral part of the approximation method. In some sense this would be a new method of approximating the centrality.

That said, if you have some justification for why or how the user should select the probability distributions for selecting the source/target nodes, this approach seems reasonable. A more general API to make this accessible could be to allow the user to provide a list of source/target values which they could generate in whatever way they prefer (presuming they know how many to provide which isn't obvious to me without looking at the code.)  

I'm not sure what you are looking for from this discussion-turned-issue. In terms of feedback on your approach, I would say it looks like a reasonable approach, but I do not believe selecting the target/source pairs will provide as good an approximation of the centrality. In terms of inclusion of this additional feature into NetworkX, my inclination is not to do so without a peer-reviewed source discussing the implications of target/source probability distributions on the approximation.

**Comment by cam4ani** [comment]
Many thanks @dschult for such a quick and helpful answer. My reason for doing this, is because the nodes correspond to habitat patches, and some are considerably smaller than others, and in turn I assume more likely to be used as a stepping stone rather than as a "destination" or "source". I am asking this question because I don't really know what their approximation relies on (e.g. how does it change with smaller or larger k etc), and therefore I was scared to be missing an important point. 

Maybe I am omitting an important step, because selecting based on probability may lead to some nodes being totally omitted from all paths used in the approximation. Though, with my assumption (i.e. nodes characterized by a smaller area would be used as stepping stone more), the betweenness values, in average, would probably still make sense biologically (e.g. that these nodes have a very low, or even 0, betweenness). But, to ensure a same and reasonable "accuracy" of the approximation across all nodes, maybe I should multiply k by 2, and use their node-pair for half of the iterations, and then my ones for the other half.
 It would help a lot to know your thoughts and most importantly criticisms on that approach, now that you know more about my assumption and reason for this. Thank you very much again for your time and help!

As a small note, I was not meaning that it should be included in the function, I totally agree with you on that (i.e., it is not the same approximation and therefore centrality measure), sorry for the confusion. It was originally meant to be on the discussion (not issue), but I turned it into an issue to increase the chance of response.
